smarts/core/coordinates.py

```diff
@@ -85,6 +86,34 @@ class Point(NamedTuple):
     y: float
     z: Optional[float] = 0
 
+    @classmethod
+    def from_np_array(cls, np_array: np.ndarray):
+        """Factory for constructing a Point object from a numpy array."""
+        assert 2 <= len(np.array) <= 3
+        z = np_array[2] if len(np_array) > 2 else 0.0
+        return cls(np_array[0], np_array[1], z)
+
+    @property
+    def as_np_array(self) -> np.ndarray:
+        """Convert this Point to a numpy array and cache the result."""
+        # Since this happens frequently and numpy array construction
+        # involves memory allocation, we include this convenience method
+        # with a cache of the result.
+        # Note that before python3.8, @cached_property was not thread safe,
+        # nor can it be used in a NamedTuple (which doesn't have a __dict__).
+        # (Points can be used by multi-threaded client code, even when
+        # SMARTS is still single-threaded, so we want to be safe here.)
+        # So we use the private global _numpy_points as a cache instead.
+        # Here we are relying on CPython's implementation of dict
+        # to be thread-safe.
+        cached = _shapely_points.get(self)
+        cached = _numpy_points.get(self)
```

Comment:
Do we need line 109, as the `cached` in line 109 is overwritten by line 110?

Reply:
Good catch.  You're right.
That was an accidental paste error.
Removed.